src/OnlineSales/Controllers/BaseControllerWithImport.cs

```diff
@@ -23,15 +23,91 @@ public BaseControllerWithImport(ApiDbContext dbContext, IMapper mapper, IOptions
     {
     }
 
+    public int ImportBatchSize { get; set; } = 5;
+
+    public long FileSizeInMB { get; set; } = 5;
+
     [HttpPost]
     [Route("import")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status422UnprocessableEntity)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status500InternalServerError)]
     public virtual async Task<ActionResult> Import([FromBody] List<TI> records)
     {
-        // TODO: Add import logic here
+        ValidateUploadFileSize();
 
-        await Task.CompletedTask;
+        var importingRecords = GetMappedRecords(records);
+
+        using (var transaction = dbContext.Database.BeginTransaction())
+        {
+            try
+            {
+                await SaveBatchChangesAsync(importingRecords);
+
+                transaction.Commit();
+            }
+            catch (Exception ex)
+            {
+                transaction.Rollback();
+                Log.Error(ex, "Error when executing batch import");
+
+                throw;
+            }
+        }
 
         return Ok();
     }
+
+    protected virtual List<T> GetMappedRecords(List<TI> records)
+    {
+        return mapper.Map<List<T>>(records);
+    }
+
+    private async Task SaveBatchChangesAsync(List<T> importingRecords)
+    {
+        int position = 0;
+
+        while (position < importingRecords.Count)
+        {
+            var batch = importingRecords.Skip(position).Take(ImportBatchSize).ToList();
+
+            foreach (var item in batch)
+            {
+                if (dbSet.Any(t => t.Id == item.Id))
+                {
+                    dbSet.Update(item);
+                }
+                else
+                {
+                    await dbSet.AddAsync(item);
+                }
+
+                position++;

```

Comment:
What is the point of "while" loop if you just make position++? Do not you need to make something like position+= ImportBatchSize?